File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/ToolImpl.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeongen.api_cache.Cache
import com.idosmart.pigeongen.api_tools.Tool
import com.idosmart.protocol_channel.innerRunOnMainThread
import com.idosmart.protocol_channel.plugin
import com.idosmart.protocol_sdk.ToolsInterface

import kotlinx.coroutines.*

internal class ToolImpl: ToolsInterface {

    private fun tool(): Tool? {
        return plugin.tool()
    }

    private fun cache(): Cache? {
        return plugin.cache()
    }

    override fun png2Bmp(
        inPath: String,
        outPath: String,
        format: Int,
        completion: (Boolean) -> Unit
    ) {
        innerRunOnMainThread {
            tool()?.png2Bmp(inPath, outPath, format.toLong()) {
                completion(it.toInt() == 0)
            }
        }
    }

    override fun compressToPNG(
        inputFilePath: String,
        outputFilePath: String,
        completion: (Boolean) -> Unit
    ) {
        innerRunOnMainThread {
            tool()?.compressToPNG(inputFilePath, outputFilePath) {
                completion(it.toInt() == 0)
            }
        }
    }

    override fun makeEpoFile(dirPath: String, epoFilePath: String, completion: (Boolean) -> Unit) {
        innerRunOnMainThread {
            tool()?.makeEpoFile(dirPath, epoFilePath, completion)
        }
    }

    override fun mp3SamplingRate(mp3FilePath: String, completion: (Int) -> Unit) {
        innerRunOnMainThread {
            tool()?.mp3SamplingRate(mp3FilePath) {
                completion(it.toInt())
            }
        }
    }

    override fun gpsInitType(motionTypeIn: Int, completion: (Boolean) -> Unit) {
        innerRunOnMainThread {
            tool()?.gpsInitType(motionTypeIn.toLong()) {
                completion(it.toInt() == 0)
            }
        }
    }

    override fun gpsAlgProcessRealtime(json: String, completion: (String) -> Unit) {
        innerRunOnMainThread {
            tool()?.gpsAlgProcessRealtime(json, completion)
        }
    }

    override fun gpsSmoothData(json: String, completion: (String) -> Unit) {
        innerRunOnMainThread {
            tool()?.gpsSmoothData(json, completion)
        }
    }

    override fun logPath(completion: (String) -> Unit) {
        innerRunOnMainThread {
            cache()?.logPath(completion)
        }
    }

    override fun currentDevicePath(completion: (String) -> Unit) {
        innerRunOnMainThread {
            cache()?.currentDevicePath(completion)
        }
    }

    override fun exportLog(completion: (String) -> Unit) {
        innerRunOnMainThread {
            cache()?.exportLog(completion)
        }
    }

    override fun lastConnectDevice(completion: (String?) -> Unit) {
        innerRunOnMainThread {
            cache()?.lastConnectDevice {
                val result = if ((it?.length ?: 0) > 2) it else null
                completion(result)
            }
        }
    }

    override fun loadDeviceExtListByDisk(sortDesc: Boolean, completion: (List<String>) -> Unit) {
        innerRunOnMainThread {
            cache()?.loadDeviceExtListByDisk(sortDesc) {
                completion(it ?: listOf())
            }
        }
    }

}

internal fun logNative(msg: String) {
    innerRunOnMainThread {
        plugin.tool()?.logNative(msg) {}
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/protocol_sdk/impl/ToolImpl.swift
//
//  ToolImpl.swift
//  protocol_channel
//
//  Created by hc on 2023/9/18.
//

import Foundation

private var _tool: Tool? {
    return SwiftProtocolChannelPlugin.shared.tools
}

private var _cache: Cache? {
    return SwiftProtocolChannelPlugin.shared.cache
}

class ToolImpl: IDOToolsInterface {
    func png2Bmp(inPath: String, outPath: String, format: Int, completion: @escaping (Bool) -> Void) {
        _runOnMainThread {
            _tool?.png2Bmp(inPath: inPath, outPath: outPath, format: format.int64, completion: { rs in
                completion(rs == 0)
            })
        }
    }
    
    func compressToPNG(inputFilePath: String, outputFilePath: String, completion: @escaping (Bool) -> Void) {
        _runOnMainThread {
            _tool?.compressToPNG(inputFilePath: inputFilePath, outputFilePath: outputFilePath, completion: { rs in
                completion(rs == 0)
            })
        }
    }
    
    func makeEpoFile(dirPath: String, epoFilePath: String, completion: @escaping (Bool) -> Void) {
        _runOnMainThread {
            _tool?.makeEpoFile(dirPath: dirPath, epoFilePath: epoFilePath, completion: { rs in
                completion(rs)
            })
        }
    }
    
    func gpsInitType(motionTypeIn: Int, completion: @escaping (Bool) -> Void) {
        _runOnMainThread {
            _tool?.gpsInitType(motionTypeIn: motionTypeIn.int64, completion: { rs in
                completion(rs == 0)
            })
        }
    }
    
    func gpsAlgProcessRealtime(json: String, completion: @escaping (String) -> Void) {
        _runOnMainThread {
            _tool?.gpsAlgProcessRealtime(json: json, completion: { rs in
                completion(rs)
            })
        }
    }
    
    func gpsSmoothData(json: String, completion: @escaping (String) -> Void) {
        _runOnMainThread {
            _tool?.gpsSmoothData(json: json, completion: { rs in
                completion(rs)
            })
        }
    }
    
    func logPath(completion: @escaping (String) -> Void) {
        _runOnMainThread {
            _cache?.logPath(completion: completion)
        }
    }
    
    func currentDevicePath(completion: @escaping (String) -> Void) {
        _runOnMainThread {
            _cache?.currentDevicePath(completion: completion)
        }
    }
    
    func exportLog(completion: @escaping (String) -> Void) {
        _runOnMainThread {
            _cache?.exportLog(completion: completion)
        }
    }
    
    func lastConnectDevice(completion: @escaping (String?) -> Void) {
        _runOnMainThread {
            _cache?.lastConnectDevice(completion: { json in
                completion(json?.count ?? 0 > 2 ? json : nil)
            })
        }
    }
    
    func loadDeviceExtListByDisk(sortDesc: Bool, completion: @escaping ([String]) -> Void) {
        _runOnMainThread {
            _cache?.loadDeviceExtListByDisk(sortDesc: sortDesc, completion: { rs in
                completion(rs ?? [])
            })
        }
    }
    
    func mp3SamplingRate(mp3FilePath: String, completion: @escaping (Int) -> Void) {
        _runOnMainThread {
            _tool?.mp3SamplingRate(mp3FilePath: mp3FilePath, completion: { rs in
                completion(rs.int)
            })
        }
    }
}


func _logNative(_ msg: String) {
    _runOnMainThread {
        SwiftProtocolChannelPlugin.shared.tools?.logNative(msg: msg, completion: { })
    }
}
